Upstream update?

orao opened the pull request:
There have been quite a few changes since you've last pulled.

iskiselev committed:
Update README.md

User should use accurev show -fi users to include in list disabled accounts (as they may have commits in repo)

Commit:
Merge pull request #32 from iskiselev/master

Unicode handling and finalize fix. Addresses issue #31.

Commit:
Merge pull request #33 from iskiselev/patch-1

Update README.md notes on listing accurev users.

orao committed:
Fixes the pop method. Issue #34.

orao committed:
Fixes the pop method. Issue #34. Forgot a part...

orao committed:
Improving the speed of the initial transition to a new branch.

orao committed:
Improving error handling when switching branches.

This commit introduces a few exceptions for situations that were
previously left ambiguous. The script should fail if it gets invalid
data back from the commandline, be it from git or accurev.

Additionally, it is expensive to simply delete everything, every time.
Though the paranoid approach of deleting everything is beneficial, and
was intentionally introduced so that you would be guaranteed to
"re-synchronize" with the AccuRev content at least every time you switch
branches, it is too expensive now that the converted git repository has
been validated to be correct.

This speeds up the branch switches by approximately 15 seconds in my
tests.

orao committed:
Enforcing invariants via exceptions.

It is important that the assumptions made through out the program hold.
On occasion I've seen strange errors that can arrise from invalid state
returned by various commands.

This commit attempts to address this by throwing exceptions and causing
the conversion process to fail when a command returns an obviously
incorrect result.

orao committed:
Preparing the stage for a new conversion method.

It is possible to iterate over the entire history and make good
decisions about where a git merge should occur. However, this requires
an entirely different approach and this commit paves the way for it to
be implemented along-side the existing methods.

Since this work is controlled via a switch it should not yet affect the
existing code so it is ok to include in the master branch.

orao committed:
Implemented the partial draft of the tr. method.

The transactions method can now handle mkstream, add and chstream
transactions. Still very much a work in progress but it should move
quickly now.

orao committed:
Continuing to improve transactions method.

Added the beginning of the implementation for the promote transaction
handling when using the transactions method.

It is still a little naive but is progressing nicely.

This also includes fixes for the accurev wrapper.

orao committed:
Fixing up debug messages.

orao committed:
Fixing initial state assumptions for process streams.

The state checks for the return values of the 'git branch' and 'git
status' commands were not correct and have caused the script to be
unable to detect the scenario where one of them returned an invalid
result.

When this happened, the processing of a stream would start from the
beginning, even though there is existing history in the repo. This
would be seen as a normal git history which at one point goes back in
time to the beginning and repeats itself.

The checks included in this commit should mitigate this issue.

orao committed:
Fixing issue #34 - pop method doesn't work.

The pop method had bitrotted while diff and deep-hist were being
developed. This fixes it but isn't ideal since the last transaction
processed should be stored instead of the last transaction that changed
this stream (which is currently the case).

orao committed:
Adding git remote command to git wrapper.

orao committed:
Implemented remote handling by the script.

Added configuration options for having remote git repositories to which
the converted streams are pushed.

orao committed:
Fixing up the handling of git remotes.

orao committed:
Moving some logic from ac2git to accurev.

orao committed:
Extract merge process into method.

orao committed:
Making affected_streams() able to check diffs.

orao committed:
Implemented the merging down into affected streams...
for the transactions method.

orao committed:
Add the new options for affected streams to the command line arguments for the accurev.py script.

orao committed:
Whitespace... sigh...

orao committed:
A failer to push to remote should not abort.

orao committed:
Refactored transaction-wise commit message format.

orao committed:
Fixing up the transactions method's merge strategy.

orao committed:
Fixing up the transactions method.

orao committed:
fixing up tr method...

orao committed:
Added defunct, purge and co commands to transactions method.

orao committed:
Fixing up the mkstream commit message.

orao committed:
Adding remote handling for transactions method.

orao committed:
Fixing the issue with unknown source streams.

It is not always possible to use `accurev hist -fex` to determine the
source stream of a transaction. In these cases the best we can do is
cherry-pick for now. It should be possible to figure out more state
later via more complex algorithms but for now the best we can do is a
cherry-pick.

orao committed:
Adding progress message to transactions method.

orao committed:
Adding ETA to transactions method.

orao committed:
Fixing potential double commit issue for transactions method. Full branch state is now stored and restored.

orao committed:
Paving the way for having include functionality in the config script... Just need to implement the merge.

orao committed:
Cleaning up the log output for transactions method.

orao committed:
Fixing uninitialized variable error in transactions method.

orao committed:
Updating the README.md and dear-contributor.md with  transactions method details.

The transactions method is now ready to move from "work in progress" to "experimental" and needs to be flagged as such in the README.md.

orao committed:
Ammend... See previous commit message.

orao committed:
Removing the ignoreTimelocks flag. Should make conversion a little quicker by default.

orao committed:
Fixing the chstream transaction processing for transactions method.

orao committed:
Merge remote-tracking branch 'origin/master'

orao committed:
Removing pytz symlink and renamind the directory. This simplifies the setup on Windows.

orao committed:
Re-adding the pytz full path name as a Linux symlink.

orao committed:
Changing the state tracking in transactions methdod from config entry to a ref to blob.

orao committed:
Making transactions method state a part of the push to remote (if any).

orao committed:
Sanitize branch names when the stream name contains spaces.

orao committed:
Fixing the cherry-pick commit message for transactions method.

orao committed:
Adding parsing of git commit output.

orao committed:
Fixing progress message. Less confusion this way.

orao committed:
Adding tracking mode option.

orao committed:
Improve message for tracking mode.

Commit:
Fixing the timezone module structure.

Commit:
Removing symlink

Commit:
Moving the egg contents to original name.

Commit:
Adding link to contents of timezone egg.

orao committed:
Added a way to recover transactions method state from the log for testing.

orao committed:
Fixing minor issue with diff when ref-names match folders...

orao committed:
Fixing up the regex expressions in git.py.

orao committed:
Fixing commit message format for the transactions method.

orao committed:
Making commit message info column aligned.

orao committed:
Basis stream should not be noted if None.

orao committed:
Add an alternative commit message style.

orao committed:
Adding more info about the recover_state_from_log.py script.

orao committed:
Stitch branches fixes.

Fixed issue #22 - windows paths in stitch_branches.sh.

With the fromStream() and toStream() functions that were added to
accurev.py recently it is now possible to make better decisions in the
branch stitching process. More specifically, a revert should no longer
cause the branch to loop back onto itself.

I've also added some preparatory work for adding source and destination
stream information to the script state information.

orao committed:
Fixed crash due to srcStream referenced before assignment.

orao committed:
Added progress info for stitch branches.

orao committed:
An empty JSON is just as invalid as no JSON, so retry.

orao committed:
Adding a sleep time for failed commands. 3 seconds.

orao committed:
Updating the caching functionality.

orao committed:
Fixing up the stitch branches filter.

orao committed:
These should have been python3 for a while now... Whoops

orao committed:
Adding the note rewriting step to stitch branches.

This allows us to continue tracking/converting a repository whose
history has been rewritten. There is one more issue to consider and that
is how to deal with a squashed commit whose notes have turned into a
list instead of a top-level object. i.e. A commit after the branch
stitching operation can represent more than one transaction state
object, but all the transaction numbers should be the same.

orao committed:
Simplifying remap_notes.py.

orao committed:
Adding source and destination stream information to state for each commit.

orao committed:
Spring cleanup and move of notes to the ac2git namespace.

orao committed:
Reworking the transactions method into a merge flag.

orao committed:
Small fixes to the new state information. Still not perfect for the merge="true" case.

orao committed:
Fixing initial condition for resuming.

orao committed:
Ammending last fix. More explicit initial condition statement.

orao committed:
Merging streams now uses the specified branch names instead of getting them from AccuRev.

orao committed:
Fixing up the initial state for merging. Now it is possible to start from any transaction.

orao committed:
Fixing exceptions.

iskiselev commented:
Thank you for note.
Really I use my fork only for PR to your master branch. I've noted some more small unicode issues, and will try to prepare PR to fix them on top of current master.

orao commented:
No problem and thanks for your help!

orao committed:
Removing incorrect invariant exception. Keeps can occur on streams (for some reverts).

orao committed:
Fixing progress message for when you don't start at transaction 1.

orao committed:
Fixing exception when processing transactions.

orao committed:
Making stream processing complete all streams at the same transaction.

orao committed:
Adding the beginning of a rewrite... Still need to do state tracking in a separate worktree.

orao committed:
Continuing to refactor the code...

orao committed:
Got the first transaction committed into the info ref.

orao committed:
Implemented accurev state information tracking.

We will track the state information needed to generate and process the
stream in a separate ref in our repo for each commit that we will be
processing later.

orao committed:
Making accurev.py more cache friendly.

orao committed:
Improving the cacheablity of the deep hist method

orao committed:
Removing command reporting.

orao committed:
Fixing the commits to hidden refs.

It seems that in a detached head state (when on a hidden ref) git
doesn't tend to update the ref with the new commit unless we explicitly
call `git update-ref <ref> <commit-hash>`. This makes sense and I should
have realised it earlier.

orao committed:
Fixing initial state assumptions.

It seams that unless we are on a branch git shows us to be in a detached
HEAD state even if we are on a hidden ref's commit chain. Additionally,
the `git status` command shows us to be detached from the first commit
in the hidden ref's history and refers to the first commit hash rather
than the ref name in the output.

My initial state assumption checks rely on the output of `git status`
giving me the branch/ref name. Since this is no longer the case the
easiest way to verify that I'm in fact on the branch I expect to be on
is to verify that the HEAD commit hash matches the hidden ref's commit
hash.

orao committed:
Implemented the data processing draft.

The processing of the accurev information which is stored in the
refs/ac2git/<depot-name>/streams/stream_<stream-number>_info ref history
is now complete and stores 3 files in each commit:
  - hist.xml
  - streams.xml
  - diff.xml

Note: the diff.xml file is omitted for the mkstream transactions.

For each stream we first get all of the above mentioned information for
the transactions that affect the stream and then do an `accurev pop` for
each of these on a separate ref,
refs/ac2git/<depot-name>/streams/stream_<stream-number>_data.

Each commit on either ref has a one line message with the following
format: 'transaction <transaction number>' which is then used as a key
into this data store or accurev state and stream contents.

The idea is to retrieve information about the transactions affecting a
stream into these two refs, meaning that we import the contents of the
stream, the output of the `accurev hist` command, the output of the
`accurev show streams` command and the output of the `accurev diff -a -i
-v <streamName> -V <streamName> -fx` command with which we can make
decisions about merging the streams at a later date.

Once everything is downloaded from accurev we will reprocess these two
refs transaction by transaction, probably in conjunction with other
streams' pairs of refs, and produce a merged output.

The expensive operation is the downloading from Accurev so adding a
stream later should be a lot cheaper, even if we have to redo the
merging, since we will already have all the necessary information.

orao committed:
Removing dead code.

orao committed:
Fixing exception

orao committed:
Fixing exception.

orao committed:
Fixing initial conditions and adding safeguards for invariants.

orao committed:
Improving logging and fixing bugs...

orao committed:
Fixup the download of the data and rename functions.

orao committed:
Rename function.

orao committed:
Removing the finalize option.

The finalize option was useful in the beginning but will be superceeded
by the feature being developed in this branch.

The idea is that choosing a merge strategy would implicitly do the post
processing of streams now that the retrieval of accurev information has
been decoupled from the processing of the branches. The new way of
post-processing is a lot cheaper and less destructive.

orao committed:
Removing finalize from the README.md. See previous commit.

orao committed:
Update README.md to have 'The result' section before the 'How it works'.

orao committed:
Remove the merge option and replace with a merge-strategy option.

orao committed:
Adding log messages for the new stages of processing.

orao committed:
Updating the README.md with the current approach information and removing no longer used processes.

orao committed:
Grammar and correctness.

orao committed:
Updating dear-contributor.md with information about the latest changes to the algorithms.

orao committed:
Remove the excess info from README.md and move it to a separate file.

The how_it_works.md is a word explanation of the algorithm while the hacking_guide.md is intended for potential developers to find their feet.

orao committed:
Updating the future.md with future work.

orao committed:
Having nothing to do is not an error. Fixing messages.

orao committed:
Fixing the commit message cleanup mode.

orao committed:
Adding more useful information about how things are done.

orao committed:
Removing remap_notes.py (required for the branch stitching that was removed earlier).

orao committed:
Refactoring copy/pasted code.

orao committed:
Refactoring repeatition away.

orao committed:
Extracting more generic code.

orao committed:
Fixing commiter being the author.

orao committed:
Fixing up the git.py script exception from last commit.

orao committed:
Implemented orphanage strategy. Existing functionality is back.

Up until this point the rewrite has broken even the basic existing
functionality which processed the streams as independent orphaned
branches in git. Finally I was able to add this back. There is still
some testing left to be done but it's getting there.

orao committed:
Fixing regressin with hidden refs.

orao committed:
Adding new commit message style to support hiding accurev information under notes.

orao committed:
Removing unneeded arguments from RetrieveStreamData.

orao committed:
Removing debug message for commit success.

orao committed:
Adding hidden ref for depots and moving streams under depotNumber instead of depotName.

In the case where we started processing a depot and it is renamed
during processing the script will have to restart if we encode the
refs with the depot name. Hence we should encode it with the depot
number instead.

This commit achieves that by maintaining a refs/ac2git/depots branch
which has the history of a single depots.xml file that is the result
of the output of the `accurev show depots -fix` command.

If the depot name requested, from the config file, is not found in
the depots.xml then we execute the accurev command and commit a new
depots.xml (only if it has changed).

The refs/ac2git/depots ref has potential merge issues when you're
using the script in a distributed manner (which is entirely possible
and intended). It will require manual merging and there's no way
around it since the `accurev show depots` command only shows you
the depots at the current point in time and is unable to provide
historical information.

orao committed:
Making the merge strategy option in the config its own field and the method able to handle the 'skip' option.

orao committed:
Fix empty commit messages hanging.

orao committed:
Fixing the accurev notes for message-style="notes"

orao committed:
Fixing the empty message commit hanging.

orao committed:
Fixing the empty message commit hanging.

orao committed:
Partially through update of the process transactions...

orao committed:
Add exception to warn of the parts that are under construction.

orao committed:
Reimplemented mkstream transaction processing for the normal merge strategy.

orao committed:
Removing useless code...

orao committed:
WIP: Reworking the promote. Outstanding issue with next_transaction storage.

orao committed:
Implemented the promote handling for the new approach.

orao committed:
Adding draft implementation for accurev add, keep, move and co operations in the new method.

orao committed:
Fixing typo.

orao committed:
Fixing state information.

orao committed:
Implemented new way of processing, chstream, defunct and purge. Consolidated processing of promote, keep, add, co, move.

orao committed:
Factorising code duplication and removing the git checkout commands. Using plumbing for rewrite.

orao committed:
Removing dead code.

orao committed:
Fix incorrect cherry-pick reports and make commit hashes 8 chars long for log messages.

orao committed:
Fix bad log message.

orao committed:
Fixing typo's in accurev.py script causing us to be unable to deremine the source stream for a promote.

orao committed:
Replacing the git checkout command with a cheaper git log command.

orao committed:
Fixing exception raised when streamMap doesn't have the source branch.

orao committed:
The source streem can never be in the list of affected streams by definition.

orao committed:
Fixing a failure to unpack.

orao committed:
Basis streams are also unlikely to be in the affected stream list... Sigh...

orao committed:
Breaks compatibility with v0.4-draft. Moving the refs, again to a different place.

orao committed:
Draft implementation of fast-forwarding.

orao committed:
TIL about packed-refs. Don't use the .git directory you idiot.

orao committed:
Clearing up the message for fast-forwarding the promotes with known source and destination streams.

orao committed:
Removing outdated warning message. Since we use this all the time we shouldn't warn about it...

orao committed:
Fixing the commit messages so that they contain src and dst streams.

orao committed:
Fixing bad fast-forwards. Should update the correct branch.

orao committed:
WIP: Rework the promote merging algorithm

The idea would be to extract the merge code from the part that
deals with the destination and source stream into a function and
call it for the affected streams which would be processed in the
order of parent to each child.

orao committed:
Adding better detection for git repositories.

orao committed:
Use rev-parse a little more wisely.

orao committed:
Testing out a refactor...

orao committed:
WIP: Implemented draft of the reworked merging strategy for child streams.

orao committed:
WIP: Removing unused code.

orao committed:
WIP: Patching up logical errors in the new code.

orao committed:
Fixed up the GetLastCommitHash() function to be able to deal with timelocks and added draft timelocks code.

orao committed:
WIP: The diff, merge-base and rev-parse commands should really be under the repo object, not the notes object.

orao committed:
WIP: Fixing minor logic errors.

orao committed:
WIP: Add children processing for a chstream.

orao committed:
WIP: Make the merging only happen on streams that don't have timelocks.

orao committed:
Merge branch 'experimental'

Reimplemented the algorithm to correctly propagate merges into affected
child streams.

orao committed:
Fixing up the mkstream message and processing. No need to perform a checkout.

orao committed:
Clarifying the messages about missing accurev information.

orao committed:
Fixing UTF-8 error.

orao committed:
Fixing more UTF-8 issues in git.py.

orao committed:
Adding more log messages for the beginning stages that can take a while to process.

orao committed:
Adding more log information for steps that can take a long time to process.

orao committed:
The push errors should be shown even if not in verbose mode.

orao committed:
Caching results of the stream searching should help find the streams later.

orao committed:
Reworked: caching results of the stream searching should help find the streams later.

orao committed:
Move the friendly message into the notes when the message style specifies notes.

orao committed:
Fixing log message.

orao committed:
Improving log message accuracy.

orao committed:
Addresses issue 42

The GetStreamMap() function was just plainly wrong. This should now
return something sensible for when no stream list is specified.

The example config was also fixed. @yfw008 pointed out that the `--`
token is invalid in an XML comment string and I used it in an example
git command (mainly `git log --notes=accurev`) which is useful to keep.
As a compromise I've removed one of the hyphens in the hope that the
user that actually reads the comment will realise this small mistake
and be able to correct it by reading the `git help log` manual.

orao committed:
Adding more detailed messages about chstream transaction processing.

orao committed:
On second thought, do not orphan the branch when we chstream to an untracked stream.

orao committed:
Fixed and simplified chstream transaction logic.

orao committed:
Adding more information for state failures.

orao committed:
Addresses issue #42.

The problem is that when we have an empty streams list we won't
retrieve the full stream-list from accurev because we were trying
to avoid adding extra calls to the accurev server. However, this
is not necessary and it is acceptable that if the user wishes to
be in `disconnected` mode w.r.t. accurev server that he is required
to have a specific stream list in the config.

orao committed:
Fixing exception to do with the return type in GetFirstTransaction(). Fixes issue #42.
